x-pack/legacy/plugins/maps/public/layers/vector_tile_layer.js

```diff
@@ -103,39 +129,59 @@ export class VectorTileLayer extends TileLayer {
       return [];
     }
     const sourceIds = Object.keys(vectorStyle.sources);
-    return sourceIds.map(sourceId => this._generateMbId(sourceId));
+    return sourceIds.map(sourceId => this._generateMbSourceId(sourceId));
   }
 
   ownsMbLayerId(mbLayerId) {
-    //todo optimize: do not create temp array
-    const mbLayerIds = this.getMbLayerIds();
-    return mbLayerIds.indexOf(mbLayerId) >= 0;
+    return mbLayerId.startsWith(this.getId());
   }
 
   ownsMbSourceId(mbSourceId) {
-    //todo optimize: do not create temp array
-    const mbSourceIds = this.getMbSourceIds();
-    return mbSourceIds.indexOf(mbSourceId) >= 0;
+    return mbSourceId.startsWith(this.getId());
   }
 
   _makeNamespacedImageId(imageId) {
     const prefix = this._source.getSpriteNamespacePrefix() + '/';
     return prefix + imageId;
   }
 
+  _requiresPrevSourceCleanup(mbMap) {
+    const sourceIdPrefix = this._generateMbSourceIdPrefix();
+    const mbStyle = mbMap.getStyle();
+    return Object.keys(mbStyle.sources).some(mbSourceId => {
+      const doesMbSourceBelongToLayer = this.ownsMbSourceId(mbSourceId);
+      const doesMbSourceBelongToSource = mbSourceId.startsWith(sourceIdPrefix);
+      return doesMbSourceBelongToLayer && !doesMbSourceBelongToSource;
```

Comment:
simple approach to check validity from the current mbMap-state 👍 Does not require additional state outside of mbMap.